CMakeLists.txt

```diff
@@ -9,6 +9,46 @@ project(opentelemetry-cpp)
 # Mark variables as used so cmake doesn't complain about them
 mark_as_advanced(CMAKE_TOOLCHAIN_FILE)
 
+if(DEFINED ENV{ARCH})
+  # Architecture may be specified via ARCH environment variable
+  set(ARCH $ENV{ARCH})
+else()
+  # Autodetection logic that populates ARCH variable
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "amd64.*|x86_64.*|AMD64.*")
+    # Windows may report AMD64 even if target is 32-bit
+    if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+      set(ARCH x64)
+    elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
+      set(ARCH x86)
+    endif()
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "i686.*|i386.*|x86.*")
+    # Windows may report x86 even if target is 64-bit
+    if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+      set(ARCH x64)
+    elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
+      set(ARCH x86)
+    endif()
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES
+         "^(aarch64.*|AARCH64.*|arm64.*|ARM64.*)")
+    set(ARCH arm64)
```

Comment:
If we are adding this, should we also be checking for 32 bit OS running on 64bit ARM, and set ARCH accordingly?

Reply:
No. The issue with unreliable arch detection is unique to Windows (`MSVC`) only.

Linux and Mac are always Okay.

I am not aware of developers building our SDK on Windows-ARM64 host machine. I am not sure even if such MSVC toolset exists. If those developers using ARM64 host as their developer box exist somewhere, then they may need to explicitly specify the `ARCH` environment variable.

Reply:
Windows ARM64 have x86/x64 MSVC run under emulation. There is no native ARM64 MSVC as for now.